src/main/drivers/serial_softserial.c

```diff
@@ -177,71 +242,131 @@ static void resetBuffers(softSerial_t *softSerial)
     softSerial->port.txBufferHead = 0;
 }
 
-serialPort_t *openSoftSerial(softSerialPortIndex_e portIndex, serialReceiveCallbackPtr rxCallback, uint32_t baud, portOptions_t options)
+serialPort_t *openSoftSerial(softSerialPortIndex_e portIndex, serialReceiveCallbackPtr rxCallback, uint32_t baud, portMode_t mode, portOptions_t options)
 {
     softSerial_t *softSerial = &(softSerialPorts[portIndex]);
 
-#ifdef USE_SOFTSERIAL1
-    if (portIndex == SOFTSERIAL1) {
-        softSerial->rxTimerHardware = &(timerHardware[SOFTSERIAL_1_TIMER_RX_HARDWARE]);
-        softSerial->txTimerHardware = &(timerHardware[SOFTSERIAL_1_TIMER_TX_HARDWARE]);
-    }
-#endif
+    int pinCfgIndex = portIndex + RESOURCE_SOFT_OFFSET;
+
+    ioTag_t tagRx = serialPinConfig()->ioTagRx[pinCfgIndex];
+    ioTag_t tagTx = serialPinConfig()->ioTagTx[pinCfgIndex];
+
+    const timerHardware_t *timerRx = timerGetByTag(tagRx, TIM_USE_ANY);
+    const timerHardware_t *timerTx = timerGetByTag(tagTx, TIM_USE_ANY);
+
+    IO_t rxIO = IOGetByTag(tagRx);
+    IO_t txIO = IOGetByTag(tagTx);
+
+    if (options & SERIAL_BIDIR) {
+        // If RX and TX pins are both assigned, we CAN use either with a timer.
+        // However, for consistency with hardware UARTs, we only use TX pin,
```

Comment:
F3 can actually use both pins in single-wire mode (probably not implemented).
But forcing user to use TX here is fine ..